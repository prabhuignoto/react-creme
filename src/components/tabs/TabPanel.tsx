import classNames from "classnames";
import React, { useMemo } from "react";

export interface TabPanelProps {
  children?: React.ReactNode;
  disabled?: boolean;
}

const TabPanel: React.FunctionComponent<TabPanelProps> = ({
  children,
  disabled,
}) => {
  const tabPanelClass = useMemo(() => {
    return classNames("rc-tab-panel", { "rc-tab-panel-disabled": disabled });
  }, []);
  return (
    <li className={tabPanelClass} role="tabpanel">
      {children}
    </li>
  );
};

export { TabPanel };
