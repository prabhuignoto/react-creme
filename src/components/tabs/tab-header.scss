@use "../../design/layout.scss";
@use "../../design/colors.scss";
@use "../../design/position.scss";
@use "../../design/icon.scss";
@use "../../design/list.scss";
@use "../../design/animate.scss";

.rc-tab-headers-wrapper {
  @extend %left;
  position: relative;
  width: 100%;

  &.rc-tab-headers-wrapper-flat::after {
    height: 3px;
  }

  &::after {
    @include position.position-abs("bottom center");
    background: colors.$secondary;
    bottom: 0;
    content: "";
    display: block;
    height: 1px;
    width: 100%;
    z-index: -1;
  }
}

.rc-tab-headers {
  @extend %list-horizontal-left;
  height: 35px;
  overflow-x: hidden;
  overflow-y: hidden;
  scroll-behavior: smooth;
  width: 90%;
  z-index: 1;

  li {
    @extend %center;
    height: 90%;
  }
}

.rc-tab-headers-control {
  @extend %center;
  height: 100%;
  justify-content: space-around;
  width: 10%;
}

.rc-tab-header-btn {
  @extend %border-radius;
  @extend %icon-md;
  cursor: pointer;

  &:hover {
    background: colors.$secondary;
  }
}

.rc-tab-header-btn-left {
  transform: rotate(180deg);
}

.rc-tab-head {
  margin: 0 0.5rem 0 0;
  position: relative;

  &.rc-tab-head-disabled {
    @extend %disabled;
  }
}

.rc-tab-header-label {
  @extend %center;
  cursor: pointer;
  height: 100%;
  min-width: 100px;
  position: relative;
  text-transform: capitalize;

  &.rc-tab-header-selected {
    color: colors.$primary;
    position: relative;
  }
}

.rc-tab-header-flat {
  &::after {
    @include position.position-abs("bottom center");
    background: rgba(colors.$primary, 70%);
    bottom: -2px;
    content: "";
    display: block;
    height: 3px;
    transition: width 0.25s cubic-bezier(0.075, 0.82, 0.165, 1);
    width: 0;
  }

  &.rc-tab-header-selected::after {
    width: 100%;
  }
}

.rc-tab-header-rounded {
  @extend %border;
  border-bottom: 0;
  border-bottom-left-radius: 0;
  border-bottom-right-radius: 0;
  position: relative;

  &.rc-tab-header-selected::after {
    @include position.position-abs("bottom center");
    background: colors.$white;
    bottom: -2px;
    content: "";
    display: block;
    height: 2px;
    width: 100%;
  }
}
