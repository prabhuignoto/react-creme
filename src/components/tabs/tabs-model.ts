import { CSSProperties, ReactNode } from "react";

export interface TabsCommonModel {
  tabStyle?: "flat" | "rounded";
  border?: boolean;
}
export interface TabHeadProps extends TabsCommonModel {
  id: string;
  name?: string;
  selected?: boolean;
  handleTabSelection: (id: string) => void;
}

export interface TabsModel extends TabsCommonModel {
  children: ReactNode[];
  labels: string[];
  width?: number | string;
  style?: CSSProperties;
}

export interface TabItemModel {
  name: string;
  id: string;
  selected?: boolean;
}

export interface TabHeadersModel extends TabsCommonModel {
  items: TabItemModel[];
  handleTabSelection: (id: string) => void;
}
