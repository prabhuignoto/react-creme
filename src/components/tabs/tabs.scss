@use "../../design/layout.scss";
@use "../../design/colors.scss";
@use "../../design/list.scss";
@use "../../design/position.scss";
@use "../../design/animate.scss";

.rc-tab-panel {
  @extend %center;
}

.rc-tabs {
  @extend %col;
  padding: 1rem;
}

.rc-tab-headers-wrapper {
  @extend %left;
  width: 100%;
}

.rc-tab-headers {
  @extend %list-horizontal-left;
  height: 35px;
  padding-top: 0.25rem;
  width: 90%;
  z-index: 0;

  li {
    @extend %center;
    height: 100%;
  }
}

.rc-tab-headers-control {
  @extend %col;
}

.rc-tab-head {
  // @include animate.halo(2px);
  margin: 0 0.5rem 0 0;
  z-index: 0;
}

.rc-tab-header-label {
  @extend %center;
  @extend %border;
  border-bottom-left-radius: 0;
  border-bottom-right-radius: 0;
  cursor: pointer;
  height: 100%;
  min-width: 100px;
  position: relative;

  &.rc-tab-header-selected {
    background: colors.$light-bg;
    color: colors.$primary;
    position: relative;
  }

  &.rc-tab-header-selected::after {
    @include position.position-abs("bottom center");
    background: colors.$light-bg;
    bottom: -2px;
    content: "";
    display: block;
    height: 2px;
    width: 100%;
  }
}

.rc-tab-panels {
  @extend %list;
  @extend %border;
  // @extend %border-radius;
  background: colors.$light-bg;
  min-height: 200px;
  width: 100%;
  z-index: -1;
}
