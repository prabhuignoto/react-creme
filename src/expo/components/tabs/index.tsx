import React from "react";
import DemoPageRenderer from "../../common/demo-page-renderer";
import Widgets from "./widgets";

function tabs() {
  return (
    <DemoPageRenderer
      tabTitles={["examples", "properties"]}
      title="Tabs"
      description="Tabs are used to group and organize content. They can be used to switch between views or to navigate between pages in a document."
      data={[
        {
          name: "tabStyle",
          description: "sets the tab style. 'flat' | 'rounded'",
          default: "flat",
          optional: "Yes",
          type: "String",
        },
        {
          name: "border",
          description: "enables or disables the border",
          default: "False",
          optional: "Yes",
          type: "Boolean",
        },
        {
          name: "labels",
          description: "sets the label for each tab",
          default: "[]",
          optional: "Yes",
          type: "Array",
        },
        {
          name: "disabledTabs",
          description: "prop to disable set of tabs",
          default: "[]",
          optional: "Yes",
          type: "Array",
        },
        {
          name: "style",
          description: "Sets any custom CSS",
          default: "{}",
          optional: "Yes",
          type: "Object",
        },
      ]}
      demoWidget={<Widgets />}
    ></DemoPageRenderer>
  );
}

export default tabs;
