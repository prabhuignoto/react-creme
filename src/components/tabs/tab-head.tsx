import classNames from "classnames";
import React, { useMemo, useRef } from "react";
import { useFocus } from "../common/effects/useFocus";
import { useKey } from "../common/effects/useKey";
import { TabHeadProps } from "./tabs-model";
import "./tabs.scss";

const TabHead: React.FC<TabHeadProps> = React.memo(
  ({
    disabled,
    focusable,
    handleTabSelection,
    id,
    name,
    selected,
    tabStyle,
  }: TabHeadProps) => {
    const ref = useRef(null);

    if (!disabled && focusable) {
      useFocus(ref, () => handleTabSelection(id));
      useKey(ref, () => handleTabSelection(id));
    }

    const headerLabelClass = useMemo(() => {
      return classNames("rc-tab-header-label", {
        "rc-tab-header-selected": selected,
        [`rc-tab-header-${tabStyle}`]: true,
      });
    }, [selected]);

    const tabHeadClass = useMemo(() => {
      return classNames("rc-tab-head", {
        "rc-tab-head-disabled": disabled,
        [`rc-tab-head-${tabStyle}`]: true,
      });
    }, [disabled]);

    return (
      <li
        key={id}
        className={tabHeadClass}
        onClick={() => !disabled && handleTabSelection(id)}
        tabIndex={!disabled && focusable ? 0 : -1}
        ref={ref}
      >
        <span className={headerLabelClass}>{name}</span>
      </li>
    );
  },
  (prev, next) =>
    prev.selected === next.selected && prev.disabled === next.disabled
);

TabHead.displayName = "TabHead";

export { TabHead };
