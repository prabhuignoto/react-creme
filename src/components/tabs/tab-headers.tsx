import classNames from "classnames";
import React, {
  startTransition,
  useCallback,
  useMemo,
  useRef,
  useState,
} from "react";
import { ChevronRightIcon } from "../../icons";
import { TabHead } from "./tab-head";
import { TabItemModel, TabsCommonModel } from "./tabs-model";
import "./tabs.scss";

export interface TabHeadersModel extends TabsCommonModel {
  items: TabItemModel[];
  handleTabSelection: (id: string) => void;
}

const TabHeaders: React.FunctionComponent<TabHeadersModel> = ({
  items,
  tabStyle,
  handleTabSelection,
}: TabHeadersModel) => {
  const tabHeadersRef = useRef<HTMLUListElement | null>(null);

  const [disableScrollRight, setDisableScrollRight] = useState(false);
  const [disableScrollLeft, setDisableScrollLeft] = useState(false);

  const [headerWidth, setHeaderWidth] = useState<number>(0);

  const canShowControls = useMemo(() => {
    if (tabHeadersRef.current) {
      const { clientWidth, scrollWidth } =
        tabHeadersRef?.current as HTMLElement;

      return clientWidth < scrollWidth;
    }
  }, [headerWidth]);

  const scrollLeft = useCallback(() => {
    const tabHeaders = tabHeadersRef.current;
    if (tabHeaders) {
      const { scrollLeft, clientWidth } = tabHeaders;
      const newLeft = Math.max(0, scrollLeft - (clientWidth * 1) / 2);
      tabHeaders.scrollLeft = newLeft;

      startTransition(() => {
        setDisableScrollLeft(newLeft === 0);
        setDisableScrollRight(false);
      });
    }
  }, []);

  const scrollRight = useCallback(() => {
    const tabHeaders = tabHeadersRef.current;
    if (tabHeaders) {
      const { scrollLeft, clientWidth, scrollWidth } = tabHeaders;
      const newLeft = scrollLeft + (clientWidth * 1) / 2;
      tabHeaders.scrollLeft = newLeft;

      startTransition(() => {
        setDisableScrollRight(newLeft + clientWidth >= scrollWidth);
        setDisableScrollLeft(false);
      });
    }
  }, []);

  const onHeadersRef = useCallback((node) => {
    tabHeadersRef.current = node;
    setHeaderWidth(node?.scrollWidth || 0);
  }, []);

  const tabHeadersWrapperClass = useMemo(() => {
    return classNames("rc-tab-headers-wrapper", {
      [`rc-tab-headers-wrapper-${tabStyle}`]: true,
    });
  }, []);

  return (
    <header className={tabHeadersWrapperClass}>
      <ul className="rc-tab-headers" ref={onHeadersRef}>
        {items.map(({ id, name, selected }) => (
          <TabHead
            key={id}
            id={id}
            name={name}
            selected={selected}
            handleTabSelection={handleTabSelection}
            tabStyle={tabStyle}
          />
        ))}
      </ul>
      {canShowControls && (
        <div className="rc-tab-headers-control">
          {!disableScrollLeft && (
            <span
              className="rc-tab-header-btn rc-tab-header-btn-left"
              role="button"
              onClick={scrollLeft}
            >
              <ChevronRightIcon />
            </span>
          )}
          {!disableScrollRight && (
            <span
              className="rc-tab-header-btn"
              role="button"
              onClick={scrollRight}
            >
              <ChevronRightIcon />
            </span>
          )}
        </div>
      )}
    </header>
  );
};

TabHeaders.displayName = "TabHeaders";

export { TabHeaders };
