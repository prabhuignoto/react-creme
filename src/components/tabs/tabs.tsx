import { nanoid } from "nanoid";
import React, { CSSProperties, useCallback, useMemo, useState } from "react";
import { TabHead } from "./tab-head";
import { TabItemModel, TabsModel } from "./tabs-model";
import "./tabs.scss";

const Tabs: React.FunctionComponent<TabsModel> = ({
  children,
  labels,
  width = 300,
}) => {
  const [items, setItems] = useState<TabItemModel[]>(
    Array.isArray(children)
      ? children.map((_, index) => ({
          id: nanoid(),
          name: labels[index],
          selected: index === 0,
        }))
      : []
  );

  const tabsStyle = useMemo(
    () =>
      ({
        "--min-width": `${width}px`,
      } as CSSProperties),
    []
  );

  const handleTabSelection = useCallback((id) => {
    setItems((prev) =>
      prev.map((item) => ({
        ...item,
        selected: id === item.id,
      }))
    );
  }, []);

  return (
    <div className="rc-tabs" style={tabsStyle} role="tab">
      <header className="rc-tab-headers-wrapper">
        <ul className="rc-tab-headers">
          {items.map(({ id, name, selected }) => (
            <TabHead
              key={id}
              id={id}
              name={name}
              selected={selected}
              handleTabSelection={handleTabSelection}
            />
          ))}
        </ul>
        <div className="rc-tab-headers-control"></div>
      </header>
      <ul className="rc-tab-panels">
        {items.map(
          ({ id, selected }, index) =>
            selected && (
              <li className="rc-tab-panel" key={id} role="tabpanel">
                {children[index]}
              </li>
            )
        )}
      </ul>
    </div>
  );
};

export { Tabs };
