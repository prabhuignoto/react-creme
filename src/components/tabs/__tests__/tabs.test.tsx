import { fireEvent, render, waitFor } from "@testing-library/react";
import React from "react";
import { act } from "react-dom/test-utils";
import { Tabs } from "../tabs";

describe("Tabs", () => {
  it("should render tabs", () => {
    const { getByRole } = render(
      <Tabs labels={["one", "two", "three"]}>
        <span>one content</span>
        <span>two content</span>
        <span>three content</span>
      </Tabs>
    );

    expect(getByRole("tablist")).toBeInTheDocument();
  });

  it("should render tab content on selection", async () => {
    const { getByText } = render(
      <Tabs labels={["one", "two", "three"]}>
        <span>one content</span>
        <span>two content</span>
        <span>three content</span>
      </Tabs>
    );

    expect(getByText("one content")).toBeInTheDocument();

    await act(async () => {
      fireEvent.click(getByText("two"));
    });

    await waitFor(async () => {
      expect(getByText("two")).toBeInTheDocument();
    });
  });

  it("should render snapshot", () => {
    const { getByRole } = render(
      <Tabs labels={["one", "two", "three"]}>
        <span>one content</span>
        <span>two content</span>
        <span>three content</span>
      </Tabs>
    );

    expect(getByRole("tablist")).toMatchSnapshot();
  });
});
