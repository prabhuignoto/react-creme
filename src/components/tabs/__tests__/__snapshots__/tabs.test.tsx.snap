// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs should render snapshot 1`] = `
<div
  class="rc-tabs"
  role="tab"
  style="--min-width: 100%;"
>
  <header
    class="rc-tab-headers-wrapper rc-tab-headers-wrapper-flat"
  >
    <ul
      class="rc-tab-headers rc-tab-headers-flat"
    >
      <li
        class="rc-tab-head rc-tab-head-flat"
        tabindex="-1"
      >
        <span
          class="rc-tab-header-label rc-tab-header-selected rc-tab-header-flat"
        >
          one
        </span>
      </li>
      <li
        class="rc-tab-head rc-tab-head-flat"
        tabindex="-1"
      >
        <span
          class="rc-tab-header-label rc-tab-header-flat"
        >
          two
        </span>
      </li>
      <li
        class="rc-tab-head rc-tab-head-flat"
        tabindex="-1"
      >
        <span
          class="rc-tab-header-label rc-tab-header-flat"
        >
          three
        </span>
      </li>
    </ul>
  </header>
  <ul
    class="rc-tab-panels"
  >
    <li
      class="rc-tab-panel"
      role="tabpanel"
    >
      <span>
        one content
      </span>
    </li>
  </ul>
</div>
`;
